runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -6710,6 +6795,123 @@ int amo_cmd_2_nic_op(fork_amo_cmd_t cmd, int fetching)
   return nic_cmd;
 }
 
+#if HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+#define MAX_BUFF_AMO_THREADS 1024
+static gni_post_descriptor_t* amo_pdc[MAX_BUFF_AMO_THREADS];
+static c_nodeid_t* amo_node_v[MAX_BUFF_AMO_THREADS];
+static int64_t* amo_vi_v[MAX_BUFF_AMO_THREADS];
+static atomic_bool* amo_lock_v[MAX_BUFF_AMO_THREADS];
+static atomic_bool amo_init_lock = false;
+static atomic_uint_least32_t amo_pdc_sz = 0;
+#endif // HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+
+static
+void buff_amo_init(void) {
+#if HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+  atomic_init_bool(&amo_init_lock, false);
+  atomic_init_uint_least32_t(&amo_pdc_sz, 0);
+#endif // HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+}
+
+void chpl_comm_atomic_buff_flush() {
+#if HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+  uint32_t sz, i;
+   sz = atomic_load_uint_least32_t(&amo_pdc_sz);
+   for(i=0; i<sz; i++) {
+    if (*amo_vi_v[i] != -1) {
+      while (atomic_exchange_bool(amo_lock_v[i], true)) { local_yield(); }
+      if (*amo_vi_v[i] != -1) {
+        post_fma_ct_and_wait(amo_node_v[i], amo_pdc[i]);
+        *amo_vi_v[i] = -1;
+      }
+      atomic_store_bool(amo_lock_v[i], false);
+    }
+  }
+#endif // HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+}
+
+static
+inline
+void do_nic_amo_nf_buff(void* opnd1, c_nodeid_t locale,
+                        void* object, size_t size,
+                        gni_fma_cmd_type_t cmd,
+                        mem_region_t* remote_mr)
+{
+#if HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+  static __thread gni_post_descriptor_t post_desc;
+  static __thread gni_ct_amo_post_descriptor_t pdc[MAX_CHAINED_AMO_LEN - 1];
+  static __thread c_nodeid_t node_v[MAX_CHAINED_AMO_LEN];
+  static __thread int64_t vi = -1;
+
+  static __thread chpl_bool inited = false;
+  static __thread atomic_bool lock;
+
+  if (inited == false) {
+    uint32_t idx;
+    while (atomic_exchange_bool(&amo_init_lock, true)) { local_yield(); }
+    atomic_init_bool(&lock, false);
```

Comment:
If you want to take a teensy bit of overhead out, `lock` can be inited outside the single-threaded region because it's private to this thread.